### PR TITLE
Add a skeleton of redesigned "pcr" package

### PR DESCRIPTION
**Pull request by xaionaro**
Adding the main entities of a new design of PCR values calculations process. New package is called "bootflow" (instead of "pcr"), since it is not longer focused on PCR values only. It also supports other roots of trust for storage and measurements.

New design (in contrast to the old one) is:
* Extensible from external packages.
* TPM terms "Event", "Extend" and "Measurement" are clarified.
* Flow now consists of steps. Each step may contain multiple measurements. It solves the problem with dynamic amount of measurements on some platforms.
* New design also aware of DICE and other chains of trust.

uml

UML is generated by command:
```
goplantuml -recursive pkg/bootflow/ | sed -e 's/\#\.\./\*--/g' | sed -re 's/\[\]([^b])/\1/g'
```

An example of a flow description:
https://github.com/9elements/converged-security-suite/pull/330/files#diff-27af1a8ff010eb439ee11da802024b19fbf9bf11ea6580efcbfed47f54135c89R9-R15

### CODE REVIEW

File: pkg/bootflow/trustchains/trust_chain_tpm.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+var _ types.TrustChain = (*TPM)(nil)
+
+type TPM struct {
+	PCRValues PCRValues
+	EventLog  TPMEventLog
+}
+
+func TPMFunc(state *types.State, fn func(tpm *TPM) error) error {
+	for _, trustChain := range state.TrustChains {
+		if tpm, ok := trustChain.(*TPM); ok {
```
Comment (mimir-d) [inline pkg/bootflow/trustchains/trust_chain_tpm.go:20]: the fact that you need to check rt type info here seems like a bit of code smell to me. you could probably keep a map of known chains and avoid this func completely; it wouldnt avoid extensibility, although realistically speaking, it's not every day that a new trust mechanism is added

Comment (mimir-d) [reply pkg/bootflow/trustchains/trust_chain_tpm.go:20]: i see the same pattern happens with state.SystemArtifacts, and you need to do a linear search every time there as well

File: pkg/bootflow/bootengine/boot_process.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/9elements/converged-security-suite/v2/pkg/bootflow/types"
+)
+
+type BootProcess struct {
+	// Do not mutate these values from the outside, the mutability
```
Comment (mimir-d) [inline pkg/bootflow/bootengine/boot_process.go:11]: you can use an accessor func here for guaranteed immutability

File: pkg/bootflow/examples/ocpdxepcr0/main.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	// the main part
+	state := types.NewState()
+	state.SetFlow(flows.OCPDXE(), 0)
+	state.EnableTrustChain(tpm.NewTPM())
+	state.EnableSystemArtifact(biosfirmware.NewBIOSFirmware(biosFirmware))
+	process := bootengine.NewBootProcess(state)
+	process.Finish()
+
+	// printing results
+	tpm.StateExec(state, func(t *tpm.TPM) error {
```
Comment (mimir-d) [inline pkg/bootflow/examples/ocpdxepcr0/main.go:34]: design nit; are there any alternatives to using this state object as an opaque pointer? this reads a bit weird to me, although it does address extensibility (which i assume was the main reason for this form in the first place)

File: pkg/bootflow/examples/bruteforce/main.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	process := bootengine.NewBootProcess(state)
+	process.Finish()
+
+	// just some debugging
+	fmt.Printf("Log:\n%#v\n", process.Log)
+
+	// == now let's bruteforce the first measurement ==
+
+	// first find the original (UNHASHED) data of the first measurement
+	tpmMeasurements := process.Log.GetVerifiedDataByTrustChainType((*tpm.TPM)(nil))
```
Comment (mimir-d) [inline pkg/bootflow/examples/bruteforce/main.go:55]: design nit; this typed nil (although valid syntax) may look weird for less experienced. is there a way to express extensibility (external i suppose in this case) without using the type system like this? id also be interested in understanding more regarding the extensibility aspect (as i see it used a lot in this whole design) and why you'd want to encode it like this (rather than, for example, leave consumers to derive their own code or contrib to this codebase directly)

File: pkg/bootflow/examples/bruteforce/main.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+	// now let's gather the rest chain (TPM init and the rest TPM Extend-s)
+	var (
+		tpmLocality *uint8
+		tpmExtends  []tpm.CommandLogEntryExtend
+	)
+	tpm.StateExec(state, func(t *tpm.TPM) error {
```
Comment (mimir-d) [inline pkg/bootflow/examples/bruteforce/main.go:62]: this linear search here to extract individual parts from the measure chain seems brittle to me and is different from the design where we had a list of tpm measurements in the sense that, in that case, we could just use the list directly without having knowledge of the flow and alter one item from it in the bruteforcer. In this case, it's basically required to recreate the flow in code (lower down in the bruterforce func) to do the same thing

File: pkg/bootflow/types/state.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+func NewState() *State {
+	return &State{
+		SystemArtifacts: map[reflect.Type]SystemArtifact{},
+		TrustChains:     map[reflect.Type]TrustChain{},
+	}
+}
+
+func (state *State) SetFlow(flow Flow, stepIdx uint) {
```
Comment (marcoguerri) [inline pkg/bootflow/types/state.go:36]: Does the state need to be rolled back when transitioning to new flow? I think we agree we it's hard to guarantee it's done correctly. Should the boot process restart from a clean state (while keeping track of the encountered issues), when a `Flow` fails and we need to switch? Heavy operations could be cached.

File: pkg/bootflow/bootengine/boot_process.go (hunk 2 above)
Comment (xaionaro) [reply pkg/bootflow/bootengine/boot_process.go:11]: Yeah, but didn't want to decide how exactly I will handle this, yet.

File: pkg/bootflow/examples/bruteforce/main.go (hunk 4 above)
Comment (xaionaro) [reply pkg/bootflow/examples/bruteforce/main.go:55]: > design nit; this typed nil (although valid syntax) may look weird for less experienced.

It looks weird even for myself :D

> is there a way to express extensibility (external i suppose in this case) without using the type system like this? 

Well. It does not require a `nil`. One may call for example:
```
tpmMeasurements := process.Log.GetVerifiedDataByTrustChainType(tpm.TPM{})
```
instead, and it still will work.

Or we can select not by type, but by a specific instance:
```
tpmInstance := tpm.NewTPM()
state.EnableTrustChain(tpmInstance)
[...]
tpmMeasurements := process.Log.GetVerifiedDataByTrustChain(tpmInstance)
```
would that be better?

File: pkg/bootflow/examples/bruteforce/main.go (hunk 5 above)
Comment (xaionaro) [reply pkg/bootflow/examples/bruteforce/main.go:62]: > in that case, we could just use the list directly without having knowledge of the flow and alter one item from it in the bruteforcer.

Here is the same. We do not know the flow, we cherry-pick the first measurement (the same as earlier) and brute force it. The differences are:
1. We manually set TPM locality, but again we extract it from the flow (and we do not need to know the flow to do that).
2. We manually do the "Extend" chain instead of calling `Calculate`. In other words, the internals of function `Calculate` are partly reimplemented in the brute-forcer.

But the brute-forcer is agnostic to the specific flow: the only requirement is to have at least the one measurement, because we brute-force it -- but it was the case in the old design as well.

File: pkg/bootflow/examples/ocpdxepcr0/main.go (hunk 3 above)
Comment (xaionaro) [reply pkg/bootflow/examples/ocpdxepcr0/main.go:34]: Could you rephrase, please?

Do I understand correctly the question is: can we avoid this `StateExec` function? If so, the answer is yes:
If above we would do:
```
tpmInstance := tpm.NewTPM()
state.EnableTrustChain(tpmInstance)
```
instead of:
```
state.EnableTrustChain(tpm.NewTPM())
```
then we could've just access `tpmInstance` directly without calling `StateExec`.

File: pkg/bootflow/actions/commonactions/set_flow.go [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+type setFlow struct {
+	nextFlowFunc   func() types.Flow
+	startStepIndex uint
+}
+
+func SetFlow(flowFunc func() types.Flow, stepIndex uint) types.Action {
+	return &setFlow{
+		nextFlowFunc:   flowFunc,
+		startStepIndex: stepIndex,
```
[the remark this answers is not in the document]
Comment (xaionaro) [reply pkg/bootflow/actions/commonactions/set_flow.go:15]: OK, good point. I added this only for extra flexibility. Just haven't thought of this.

Yeah, to remove `stepIndex` is easy (actually easier than preserve it).

File: pkg/bootflow/bootengine/boot_process.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+		return nil, nil, nil, false
+	}
+
+	step := state.CurrentFlow[state.CurrentStepIdx]
+	actions := step.Actions(state)
+	var stepIssues StepIssues
+	for idx, action := range actions {
+		issue := action.Apply(state)
+		if issue != nil {
+			stepIssues = append(stepIssues, StepIssue{ActionIndex: uint(idx), Issue: issue})
```
[the remark this answers is not in the document]
Comment (xaionaro) [reply pkg/bootflow/bootengine/boot_process.go:38]: > Also `ActionIndex` tracking for issues seems to be very fragile.

`ActionIndex` is supposed to be used only for debugging. The specific Issue should be defined by error type returned by `action.Apply`.

File: pkg/bootflow/types/state.go (hunk 6 above)
Comment (xaionaro) [reply pkg/bootflow/types/state.go:36]: > Does the state need to be rolled back when transitioning to new flow?

In the current proposal we do not roll back.

> Should the boot process restart from a clean state (while keeping track of the encountered issues), when a `Flow` fails and we need to switch?

In the current proposal we are not restarting the boot process.

---

The `SetFlow` is like `JMP`, not like `CALL`. There is no returning to the original `Flow`, no saving state. Just sequentially execute steps until there will be no more steps to execute.

Basically a `Flow` might be of the of, for example (in pseudo-code):
```
Root {
    IfFITPresent{SetFlow{Intel}}
    IfAMDManifestPresent{SetFlow{AMD}}
}

Intel {
    IfBPMPresent{SetFlow{CBnT}}
    SetFlow{LegacyTXT}
}

CBnT {
    ValidateACM{Failure: SetFlow{CBnTFailure}}
    ValidateKM{Failure: SetFlow{CBnTFailure}}
    ValidateBPM{Failure: SetFlow{CBnTFailure}}
    ValidateIBB{Failure: SetFlow{CBnTFailure}}
    TPMInit(3)
    MeasurePCR0DATA
    SetFlow{IntelResetVector}
}

CBnTFailure {
    IfFITAddrIs{specialFitType, specialAddrValue, CBnT0TFailureMeta} // check if a special FIT entry has a special value
    SetFlow{IntelResetVector}
}

CBnT0TFailureMeta {
    MeasurePCR0DATAFailure
    SetFlow{IntelResetVector}
}

IntelResetVector {
    IfOCP{SetFlow{OCPPEI}}
    IfHPE{SetFlow{HPEPEI}}
    IfLenovo{SetFlow{LenovoPEI}}
}

OCPPEI {
    TPMInitLazy(3)
    MeasureUEFIGUID(0, OCP_PEI_GUID)
    MeasureSeparator(0)
    SetFlow{OCPDXE}
}

OCPDXE {
    MeasurePCDVariable(0, "FirmwareVendorVersion")
    MeasureUEFIGUID(0, OCP_DXE_GUID)
    MeasureSeparator
}
```

Comment (marcoguerri) [reply pkg/bootflow/types/state.go:36]: Ok, as long as the contract is clear that a flow should be built to leave consistent state behind. Hard to enforce, but this is what should happen in real world. It's not always the case, but this doesn't mean we should do it right here :) 

As an example, Intel ACM I think behaves slightly differently (e.g. requirement to cap PCR registers after initializing TPM if NVRAM indexes are provisioned in wrong way). Intel ACM would actually have:

```
    TPMInit(3)
    ValidateACM{Failure: SetFlow{CBnTFailure}}
    ValidateKM{Failure: SetFlow{CBnTFailure}}
    ValidateBPM{Failure: SetFlow{CBnTFailure}}
    ValidateIBB{Failure: SetFlow{CBnTFailure}}
    MeasurePCR0DATA
    SetFlow{IntelResetVector}
```

But ok, Intel ACM is the broken one, no need to replicate broken behavior :)

File: pkg/bootflow/trustchains/trust_chain_tpm.go (hunk 1 above)
Comment (xaionaro) [reply pkg/bootflow/trustchains/trust_chain_tpm.go]: Fixed. Now we use a map instead

File: pkg/bootflow/bootengine/boot_process.go (hunk 2 above)
Comment (mimir-d) [reply pkg/bootflow/bootengine/boot_process.go:11]: ok, just making sure it's not left as a "temporary todo" :D

File: pkg/bootflow/examples/ocpdxepcr0/main.go (hunk 3 above)
Comment (mimir-d) [reply pkg/bootflow/examples/ocpdxepcr0/main.go:34]: oh... so that's the semantic. Wasnt clear to me from the naming "StateExec". Well, why not have a "GetChain" instead of the lambda call? What's the benefit here? (i think you have a similar pattern for Log)

File: pkg/bootflow/examples/bruteforce/main.go (hunk 4 above)
Comment (mimir-d) [reply pkg/bootflow/examples/bruteforce/main.go:55]: I think this is an artifact of the design. You're essentially using the type system as an enum value here (and discarding the actual instance value, which here is nil, or even worse tpm.TPM{}). It's usually more clear to be direct about it (as in, use an actual enum), but then again, I don't have a simple localized proposal for this right now considering all the other code

File: pkg/bootflow/examples/bruteforce/main.go (hunk 5 above)
Comment (mimir-d) [reply pkg/bootflow/examples/bruteforce/main.go:62]: The bruteforcer is agnostic, yes. But this call still implies knowledge about the flow by virtue of having to know that there is a TPM locality initializer in it (or in a non-trivial example, some other list of measurements). It still seems to me that you have a mirror of the flow in the bruteforce callback func, in order to compute the final hash to compare, whereas in the old design, the Calculate method dealt with this impl aspect, rather than the bruteforcer callback impl (and the Calculate encapsulated the flow logic).
Then again, maybe im misinterpreting something.

File: pkg/bootflow/examples/ocpdxepcr0/main.go (hunk 3 above)
Comment (xaionaro) [reply pkg/bootflow/examples/ocpdxepcr0/main.go:34]: It was just for syntax sugar.

Talked with @mimir-d in a video chat. Decided to go with `GetChain` instead.

File: pkg/bootflow/examples/bruteforce/main.go (hunk 5 above)
Comment (xaionaro) [reply pkg/bootflow/examples/bruteforce/main.go:62]: Talked with @mimir-d: we decided to reuse the TPM Extend logic instead of duplicating it.

File: pkg/bootflow/bootengine/boot_process.go (hunk 2 above)
Comment (xaionaro) [reply pkg/bootflow/bootengine/boot_process.go:11]: Frankly, these details does not seem a priority any soon :)
There are downsides of using an accessor. For example you cannot basically serialize a structure, which does not expose fields. I'm yet not sure how exactly I will handle this.

File: pkg/bootflow/types/state.go (hunk 6 above)
Comment (xaionaro) [reply pkg/bootflow/types/state.go:36]: > Intel ACM would actually have:

I'm actually not sure. It actually might be than `TPMInit(3)` happens only on on-need basis (not earlier).

File: pkg/bootflow/types/flow.go [hunk 9]
```diff
@@ -0,0 +1,29 @@
+package types
+
+// Flow describes steps of the boot process.
+//
+// A flow is static (never change).
+type Flow []Step
+
+// Stop describes a single step of a boot process, essential for the measurements.
```
Comment (rihter007) [inline pkg/bootflow/types/flow.go:8]: Stop -> Step?

File: pkg/bootflow/types/state.go (hunk 6 above)
Comment (rihter007) [reply pkg/bootflow/types/state.go:36]: Hm... that looks counter intitive. So different flows should know about each other + that redefinition of the flow for the step.

I would propose to make Flow an interface that returns the next step, this will free flows to know about each other, only share the steps (which they are doing anyway)

```
  type Flow []Step
  
  type Flow interface {
       Next() (Step, error) // returns nil if this is the last step or error is smth went wrong and the next step could not be determined
  }
```

File: pkg/bootflow/types/trust_chain.go [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+type TrustChains map[reflect.Type]TrustChain
+
+type TrustChain interface {
```
Comment (rihter007) [inline pkg/bootflow/types/trust_chain.go:11]: Could you please provide a couple of examples?
It looks like that we need more methods, to at least check for error

And why do we need IsInitialized?

File: pkg/bootflow/types/system_artifact.go [hunk 11]
```diff
@@ -0,0 +1,28 @@
+package types
+
+import (
+	"fmt"
+	"io"
+	"reflect"
+	"strings"
+)
+
+type SystemArtifacts map[reflect.Type]SystemArtifact
```
Comment (rihter007) [inline pkg/bootflow/types/system_artifact.go:10]: Maybe let's use an ID rather then reflect.Type?
This will allow to share code of "firmware part" artifact which all will share same functionality, but just refer to logically different pieces of firmware

File: pkg/bootflow/types/state.go [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+	key := typeMapKey(systemArtifactSample)
+	systemArtifact := state.SystemArtifacts[key]
+	if systemArtifact == nil {
+		return ErrNoSystemArtifact{SystemArtifactKey: key}
+	}
+
+	return callback(systemArtifact)
+}
+
+func (state *State) EnableSystemArtifact(systemArtifact SystemArtifact) {
```
Comment (rihter007) [inline pkg/bootflow/types/state.go:79]: I would call AddSystemArtifact :)

File: pkg/bootflow/types/system_artifact.go (hunk 11 above)
Comment (rihter007) [reply pkg/bootflow/types/system_artifact.go:10]: + what if some steps will share artifacts? (but say have different types of checks or even similar, that can happen in our abstract world)

File: pkg/bootflow/types/state.go [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+	"reflect"
+	"strings"
+)
+
+// State describes a virtual state of a machine being booted
+type State struct {
+	// Do not mutate these values from the outside, the mutability
+	// is owned by State itself.
+
+	SystemArtifacts SystemArtifacts
```
Comment (rihter007) [inline pkg/bootflow/types/state.go:14]: Will we store somewhere the result of check of an artifact? Is it a TrustChain? Can TrustChain use each other?

Comment (rihter007) [reply pkg/bootflow/types/state.go:14]: Ah sorry, seems like SystemArtifact is firmware, while pieces of it that we verify are in VerifiedData?

File: pkg/bootflow/systemartifacts/biosfirmware/bios_firmware.go [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+		return fn(artifact.(*BIOSFirmware))
+	})
+}
+
+func (fw *BIOSFirmware) Size() uint {
+	return uint(len(fw.Content))
+}
+
+func (fw *BIOSFirmware) ReadAt(b []byte, offset int64) (n int, err error) {
+	return bytes.NewReader(fw.Content).ReadAt(b, offset)
```
Comment (rihter007) [inline pkg/bootflow/systemartifacts/biosfirmware/bios_firmware.go:34]: Some our interfaces operate on bytes :(

File: pkg/bootflow/types/trust_chain.go (hunk 10 above)
Comment (rihter007) [reply pkg/bootflow/types/trust_chain.go:11]: Found examples, sorry :)

File: pkg/bootflow/systemartifacts/biosfirmware/bios_firmware.go (hunk 14 above)
Comment (xaionaro) [reply pkg/bootflow/systemartifacts/biosfirmware/bios_firmware.go:34]: It will be my job to fix that :)

**Comment by ChriMarMe** [comment]
This got closed due to change of default branch from master to main. Rebase on main and rename branch, open PR again is a legit move here.

File: pkg/bootflow/examples/ocpdxepcr0/main.go (hunk 3 above)
Comment (xaionaro) [reply pkg/bootflow/examples/ocpdxepcr0/main.go:34]: Changed my mind. Since now we migrated CSS to go1.18, now we can use generics, so in the end I'm going with:
```
tpmInstance := tpm.GetFrom(state)
```
instead of:
```
tpm.StateExec(state, func(tpmInstance *tpm.TPM) error {
```

File: pkg/bootflow/examples/bruteforce/main.go (hunk 4 above)
Comment (xaionaro) [reply pkg/bootflow/examples/bruteforce/main.go:55]: Changed to:
```
tpmMeasurements := process.Log.GetDataVerifiedBy(tpmInstance)
```

File: pkg/bootflow/examples/bruteforce/main.go (hunk 5 above)
Comment (xaionaro) [reply pkg/bootflow/examples/bruteforce/main.go:62]: Done. Here is the example: https://github.com/9elements/converged-security-suite/blob/a22e7ec0c568acf6c45271250efa9f9ce2827c0e/pkg/bootflow/examples/bruteforce/main.go

File: pkg/bootflow/actions/commonactions/set_flow.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (xaionaro) [reply pkg/bootflow/actions/commonactions/set_flow.go:15]: Fixed by 60c0e0b6aa636851b4a7c0a398d47202383705d8

File: pkg/bootflow/types/state.go (hunk 6 above)
Comment (xaionaro) [reply pkg/bootflow/types/state.go:36]: > Ok, as long as the contract is clear that a flow should be built to leave consistent state behind. 

One of the major idea of this new design: to be close to how actual boot process works (to be able to reproduce its behavior). And unfortunately in the actual boot process we have no such guarantees, thus in these packages not having this guarantees is considered a feature (not a bug): if something in a real boot leaves an unclean state, we want to have exactly the same thing here.

> So different flows should know about each other + that redefinition of the flow for the step.

In reality (in a real boot process) they do know about entry points of each other.

> I would propose to make Flow an interface that returns the next step, this will free flows to know about each other, only share the steps (which they are doing anyway)

Sorry, I do not understand. They will still to know about each other to be able to jump to each other. Otherwise we need to denormalize the pseudo-code tree given above (and make it a bunch of flat sequential steps instead of being a tree).

File: pkg/bootflow/types/flow.go [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+// Steps of a flow may vary depending on a State.
+//
+// An example: measure specific sections in an AMD Manifest
+type Step interface {
+	Actions(*State) Actions
```
[the remark this answers is not in the document]
Comment (xaionaro) [reply pkg/bootflow/types/flow.go:13]: There are two semantic layers:
* What to do.
* How to do.

An example of "what to do" is "measure DXE", and corresponding example of "how to do" will be "find UEFI volume, hash it and use TPMExtend and TPMEventLogAdd". Here:
* "Measure DXE" is a `Step`.
* "find UEFI volume [...]" is `Action` (/`Actions`)

For example as you might remember in AMD we have a semantically single step to measure some manifest data, but the implementation requires actually of lot of measurements. So a `Flow` always has constant list of `Step`-s, but a `Step` consts if a list of `Actions` which depends on the State (of the imaginary machine).

File: pkg/bootflow/types/trust_chain.go (hunk 10 above)
Comment (xaionaro) [reply pkg/bootflow/types/trust_chain.go:11]: > Could you please provide a couple of examples?

Here is an example of using a `TrustChain`:
https://github.com/9elements/converged-security-suite/blob/60c0e0b6aa636851b4a7c0a398d47202383705d8/pkg/bootflow/examples/ocppeipcr0/main.go

There you can find:
```go
	tpmInstance, err := tpm.GetFrom(state)
	if err != nil {
		panic(err)
	}
```

It extracts `TrustChain` with type `*TPM` from the `state`.

> It looks like that we need more methods, to at least check for error

Unfortunately, different `TrustChain`-s may have different designs (for example DICE and TPM measurements) and because of that we do not have common methods.
But yet we need an interface we will use for the `map` of all `TrustChain`-s.

> And why do we need `IsInitialized`?

That was the only common method I thought of for different `TrustChain`-s. So two reasons:
1. One may use it to know if this `TrustChain` was initialized during a given boot process. If are more common methods for any abstract `TrustChain`, I'd gladly add them here to unlock more opportunities.
2. To do not have `interface{}` and thus to enforce compile time checks that a good type is provided in an argument to somewhere.

File: pkg/bootflow/types/state.go (hunk 12 above)
Comment (xaionaro) [reply pkg/bootflow/types/state.go:79]: The problem with `Add` is when we "add apple, and then again add apple" we get "two apples". While here it is not the case: here "each unique apple can be presented only once".

Renamed to `Include`. Hoping now is better :)

File: pkg/bootflow/types/state.go (hunk 13 above)
Comment (xaionaro) [reply pkg/bootflow/types/state.go:14]: > SystemArtifact is firmware, while pieces of it that we verify are in VerifiedData?

Yes. Though not only "firmware" might be `SystemArtifact`. For example "status registers" is also a `SystemArtifact`.

File: pkg/bootflow/types/flow.go (hunk 9 above)
Comment (xaionaro) [reply pkg/bootflow/types/flow.go:8]: Fixed.
